src/Lecture3.hs

```diff
@@ -122,19 +133,24 @@ data Reward = Reward
     } deriving (Show, Eq)
 
 instance Semigroup Reward where
+  (<>) (Reward a True) (Reward b _) = Reward (a <> b) True
+  (<>) (Reward a _) (Reward b True) = Reward (a <> b) True
+  (<>) (Reward a _) (Reward b _) = Reward (a <> b) False
 
 
 instance Monoid Reward where
+  mempty = Reward (Gold 0) False
 
 
 {- | 'List1' is a list that contains at least one element.
 -}
+
 data List1 a = List1 a [a]
     deriving (Show, Eq)
 
 -- | This should be list append.
 instance Semigroup (List1 a) where
-
+  (<>) (List1 a1 a2) (List1 b1 b2) = List1 a1 (a2 <> [b1] <> b2)
```

Comment:
You can slightly improve the performance here by using the `:` operator instead of list append `<>` or `++` 🐎 
```suggestion
  List1 x xs <> List1 y ys = List1 x (xs <> (y : ys))
```